src/main/java/org/candlepin/resource/util/EntitlementEnvironmentFilter.java

```diff
@@ -0,0 +1,147 @@
+import java.util.List;
+import java.util.Set;
+
+/*
+ * The EntitlementEnvironmentFilter class filters the consumers entitlement who's entitlement certificates are
+ * needed to be re-generated because of changes (addition, removal or re-prioritization) in
+ * consumer existing environment. This filtering avoids re-generating all of the
+ * consumer entitlement certificates.
+ */
+public class EntitlementEnvironmentFilter {
+    private EntitlementCurator entitlementCurator;
+    private EnvironmentCurator environmentCurator;
+    private List<String> preExistingEnvs;
+    private List<String> updatedEnvs;
+    private Consumer consumer;
+
+    public EntitlementEnvironmentFilter(EntitlementCurator entitlementCurator,
+        EnvironmentCurator environmentCurator) {
+        this.entitlementCurator = entitlementCurator;
+        this.environmentCurator = environmentCurator;
+    }
+
+    public EntitlementEnvironmentFilter setPreExistingEnvironments(List<String> preExistingEnvs) {
+        this.preExistingEnvs = preExistingEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setUpdatedEnvironment(List<String> updatedEnvs) {
+        this.updatedEnvs = updatedEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setConsumerToBeUpdated(Consumer consumer) {
+        this.consumer = consumer;
+        return this;
+    }
+
+    /**
+     * Returns the list of environment IDs which got added, removed or re-prioritized
+     * from consumers current existing environments.
+     *
+     * @return
+     *  Returns the unique environmentIds which got added, removed or re-prioritized
+     */
+    private List<String> getEnvironmentChangeList() {
+        Set<String> envChangeList = new HashSet<>();
+
+        // Envs Added
+        envChangeList.addAll(CollectionUtils.subtract(this.updatedEnvs, this.preExistingEnvs));
+
+        // Envs removed
+        envChangeList.addAll(CollectionUtils.subtract(this.preExistingEnvs, this.updatedEnvs));
+
+        // Envs reordered (Index within list changes)
+        for (String env : this.updatedEnvs) {
+            if (this.preExistingEnvs.indexOf(env) != this.updatedEnvs.indexOf(env)) {
+                envChangeList.add(env);
+            }
+        }
+
+        return new ArrayList<>(envChangeList);
+    }
+
+    /**
+     * Filters the entitlements who's certificates are needed to be re-generated due to environments
+     * being added, removed or re-prioritized for a consumer.
+     * Idea is to avoid regenerating all of the consumer's entitlements certs & filter the
+     * necessary entitlements who's certificates are actually needed to be re-generated.
+     *
+     * @return
+     *  List of entitlement selected for certificate regeneration
+     */
+    public List<Entitlement> filterEntitlements() {
+        List<Entitlement> entitlements = new ArrayList<>();
+        List<String> environmentChangeList = getEnvironmentChangeList();
```

Comment:
Also, I know we discussed this offline, but just adding a note for myself when I re-review after the break: We want to abstract this filtering functionality (and the two queries that will be made in total) a step further, by accepting multiple consumers (as opposed to performing 2 queries per consumer, which would be another N+1 problem case).

I'd say you can change the APIs on this class as you see the best fit for this